### PR TITLE
Feat: Generate typescript package types and tests

### PR DESCRIPTION
**Pull request by NickBolles**
Resubmitting as a pr to the base repo. 

Original PR: https://github.com/ChangJoo-Park/create-nuxt-app/pull/1
But it didn't make it into the initial typescript support here: https://github.com/nuxt/create-nuxt-app/pull/449

## List of changes 
1. Create nuxt.config.ts
2. Update nuxt.config.ts with changes from nuxt.config.js
3. Create a generate component template helper to switch between javascript and typescript

    - Another approach, if there are more typscript differences, could be a seperate file for js and typescript. but that seemed unnecessary right now

4. Update the following components to generate typescript components using the generateComponent template helper
    - buefy
    - FrameVuerk
    - Iview
    - Vuesax
    - Vuetify
5. Add typescript files to eslint and lint-staged configs
6. fix tests for typescript runtime
7. Add vue-shims.d.ts file

### Issues from #448 - @davelsan

1. [Fix] File XXX is not a module (Vetur error)
 - vue-shims.d.ts included above
2. [Fix] Remove linting of unused vars
  - Add a fix by prefixing them with a `_`, which is the standard way to declare an unused var
3. [Fix] Do not use the base ESLint indent rule
 - No fix for this in this PR. I'm going to leave this for another issue
4. [Feat] Add TypeScript linting to package.json scripts
- Included above
5. [Feat] Add TypeScript support to Jest test files
- added @types/jest as suggested by @davelsan
6. [Feat] Port Nuxt config to TypeScript
- Included above

### CODE REVIEW

**Comment by mtltechtemp** [comment]
@NickBolles Ping ;)

**Comment by NickBolles** [comment]
I'll try to look at this again today or tomorrow. Ping me again if you don't hear back by monday

**Comment by NickBolles** [comment]
Issues from #448 

1. [Fix] File XXX is not a module (Vetur error)
 - vue-shims.d.ts already included in this PR
2. [Fix] Remove linting of unused vars
  - Add a fix by prefixing them with a `_`, which is the standard way to declare an unused var
3. [Fix] Do not use the base ESLint indent rule
 - No fix for this in this PR (yet). I don't quite understand the issue yet.
4. [Feat] Add TypeScript linting to package.json scripts
- Already included in this PR
5. [Feat] Add TypeScript support to Jest test files
- added
6. [Feat] Port Nuxt config to TypeScript
- Already included in this PR

todo: look through the rest of the modules to see if we can install types for them

**Comment by davelsan** [comment]
Thanks for the new additions. A couple of notes:


#### [Fix] File XXX is not a module (Vetur error)
> vue-shims.d.ts already included in this PR

This should be enough. For the additional problem reported in vuejs/vetur#1187, there is a PR vuejs/vetur#1806 on the way.

#### [Fix] Do not use the base ESLint indent rule
> - No fix for this in this PR (yet). I don't quite understand the issue yet.

Originally reported in nuxt/eslint-config#76.

The issue is that, within `*.vue` components, indentation enforcement should be handled by `eslint-plugin-vue` through the [`vue/script-indent`](https://github.com/vuejs/eslint-plugin-vue/blob/master/docs/rules/script-indent.md) rule.

In the docs (see above link) it is recommended that the base `indent` rule is switched off within `*.vue` files, using an override.

```jsonc
{
  "rules": {
    "vue/script-indent": ["error", 2, { "baseIndent": 1 }]
  },
  "overrides": [
    {
      "files": ["*.vue"],
      "rules": {
        "indent": "off"
      }
    }
  ]
}
```

In `*.ts` files indentation should enforced by the [`@typescript-eslint/indent`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/indent.md) rule. In the docs it is recommended to switch off the base `indent` rule.

 ```jsonc
{
  "rules": {
    "indent": "off",
    "vue/script-indent": ["error", 2, { "baseIndent": 1 }]
  },
  "overrides": [
    {
      "files": ["*.ts"],
      "rules": {
        "@typescript-eslint/indent": ["error", 2]
      }
    }
  ]
}
```

**Comment by NickBolles** [comment]
> This should be enough. For the additional problem reported in [vuejs/vetur#1187](https://github.com/vuejs/vetur/issues/1187), there is a PR [vuejs/vetur#1806](https://github.com/vuejs/vetur/pull/1806) on the way.

Nice! This should be good for now though. 



> #### [Fix] Do not use the base ESLint indent rule
> > * No fix for this in this PR (yet). I don't quite understand the issue yet.

Your reasoning sounds good to me.

Could we do something like this though? 
```
{
  "rules": {
    "vue/script-indent": ["error", 2, { "baseIndent": 1 }]
  },
  "overrides": [
    {
      "files": ["*.vue","*.ts"],
      "rules": {
        "indent": "off"
      }
    }
  ]
}
```

Would that result in the same functionality and also retain the "indent" rule for non ts or vue files?

**Comment by davelsan** [comment]
> Would that result in the same functionality and also retain the "indent" rule for non ts or vue files?

Yes, that would work, but then we are not using the `@typescript-eslint/indent` rule for `*.ts` files, and it feels weird to have a `vue-eslint/indent` base config with a `*.vue` override as well.

To avoid conflicts between rules, we could decide on a base config and turn on/off rules for specific files.

For a Nuxt-TypeScript project, I'd suggest the following:

- Base: set `vue/script-indent` and turn off `indent`
- Override: set `@typescript-eslint/indent rule` for `*.ts` files
- (Optional) Override: set `indent` for `*.js` files

Because all three rules conflict with each other, I think this is a more parsimonious approach.

```jsonc
{
  // .eslintrc.json
  "rules": {
    "indent":"off",
    "vue/script-indent": [ "error", 2, {
      "baseIndent": 1
    }]
  },
  "overrides": [
    {
      "files": [ "*.ts" ],
      "rules": {
        "@typescript-eslint/indent": ["error", 2]
      }
    },
    {
      "files": [ "*.js" ],
      "rules": {
        "indent": [ "error", 2 ]
      }
    }
  ]
}
```

If you prefer to have `indent` on by default (I don't recommend this), then `vue/script-indent` and `@typescript-eslint/indent` should be in separate overrides, plus `indent` should be switched off in each.

---

This is just a side note. For reasons unknown to me yet, even in `*.js` files the `indent` rule causes the following error in `vscode` (e.g. when typing `const`)

> ESLint: Cannot read property 'loc' of undefined

Maybe this is a problem that should be reported in the [`microsoft/vscode-eslint`](https://github.com/Microsoft/vscode-eslint) repo.

**Comment by NickBolles** [comment]
i'm not sure why there are so many commits. I had rebased onto master, not picking all of ChangJoo Parks commits. Diffs are only from my commits though.

@kevinmarrec could you review?

**Comment by mtltechtemp** [comment]
@NickBolles Might be related to this issue https://github.com/vuejs/eslint-plugin-vue/issues/370

**Comment by NickBolles** [comment]
> @NickBolles Might be related to this issue https://github.com/vuejs/eslint-plugin-vue/issues/370

That's the eslint error. But it's not the same cause. The cause is that we are using ejs tags (`<%_ _%>` for example) in Vue files. That's unexpected in most cases. 

Linting for anything with ejs syntax seems like a challenge. 

That being said, I think disabling that rule for the repo would be the fix.

**Comment by davelsan** [comment]
Hey, I wanted to drop by and say I appreciate _very much_ the work you're doing in this PR. Sometimes the way I write is not the most sensible, so I'm just trying to make sure I convey the right message.

So, yeah, looking forwards to this being merged!

> That being said, I think disabling that rule for the repo would be the fix.

(Trying to stay on topic)  I don't know the next thing about `SAO` or `ejs`, but my uneducated guess is that you're right. I say this after various messy and unsuccessful attempts to generate the `<script/>` tags without the linter yelling at me.

**Comment by NickBolles** [comment]
> Thanks for the pr, could you please fix the lint errors ?

done!

**Comment by mtltechtemp** [comment]
@NickBolles Is this ready now?

**Comment by mtltechtemp** [comment]
@davelsan and @NickBolles You might find this useful https://github.com/overlookmotel/eslint-plugin-ejs-js

**Comment by NickBolles** [comment]
Is this important enough to patch the package? Seems like it might belong in a seperate pr?

**Comment by mtltechtemp** [comment]
> Is this important enough to patch the package? Seems like it might belong in a seperate pr?

That's just a 3 line patch to make the ejs lib do the work.

**Comment by mtltechtemp** [comment]
@NickBolles Did you understand why we have `error clear vue/comment-directive` ?

**Comment by mtltechtemp** [comment]
@NickBolles ping

**Comment by NickBolles** [comment]
@mtltechtemp If you want to put a pull request together, either to follow up this one, or to merge into this branch I'd be fine with that. Although IMO a better fix would be to update eslint-plugin-ejs to register a pre-processor ID (https://eslint.org/docs/developer-guide/working-with-plugins#processors-in-plugins) and then allow users to configure it via the processor config, or overrides configs: https://eslint.org/docs/user-guide/configuring#specifying-processor

I hate doing one time patches of modules when with a little bit more work everyone can benefit.

File: template/frameworks/vuesax/pages/index.vue [hunk 1]
```diff
@@ -61,6 +61,15 @@
   </div>
 </template>
 
+<%- generateComponent(`
```
Comment (kevinmarrec) [inline template/frameworks/vuesax/pages/index.vue:64]: Sounds like the `script` section hasn't been removed from this file

File: template/nuxt/nuxt.config.js [hunk 2]
```diff
@@ -86,6 +86,7 @@ module.exports = {
   /*
   ** Nuxt.js dev-modules
   */
+
```
Comment (kevinmarrec) [inline template/nuxt/nuxt.config.js:89]: Don't think this carriage return is needed as it's not done for the other comments on top of properties.
```suggestion
```

File: template/nuxt/nuxt.config.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    <%_ } else if (ui === 'framevuerk') { _%>
+    '@/plugins/framevuerk'
+    <%_ } else if (ui === 'vuesax') { _%>
+    '@/plugins/vuesax'
+    <%_ } _%>
+  ],
+  /*
+  ** Nuxt.js dev-modules
+  */
+
```
Comment (kevinmarrec) [inline template/nuxt/nuxt.config.ts:82]: Don't think this carriage return is needed as it's not done for the other comments on top of properties.
```suggestion
```

File: template/nuxt/nuxt.config.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
```
Comment (kevinmarrec) [inline template/nuxt/nuxt.config.ts:1]: Useless carriage return at beginning of file
```suggestion
```

File: template/tsconfig.json [hunk 5]
```diff
@@ -25,7 +25,9 @@
     },
     "types": [
       "@types/node",
-      "@nuxt/types"
+      "@nuxt/types"<%_ if (ui === 'vuetify') { _%>,
+        "vuetify"
```
Comment (kevinmarrec) [inline template/tsconfig.json:29]:
```suggestion
        "@nuxtjs/vuetify"
```

File: test/index.test.js [hunk 6]
```diff
@@ -28,9 +28,16 @@ const verifyPkg = async (t, answers = {}) => {
...
   const stream = await sao.mock({ generator }, answers)
-  const configFile = answers.server === 'adonis' ? 'config/nuxt.js' : 'nuxt.config.js'
+  let configFile
+  if (answers.server === 'adonis') {
+    configFile = 'config/nuxt.js'
+  } else if (typescript) {
+    configFile = 'nuxt.config.ts'
+  } else {
+    configFile = 'nuxt.config.js'
+  }
```
Comment (kevinmarrec) [inline test/index.test.js:40]:
```suggestion
  const configFile = answers.server === 'adonis' ? 'config/nuxt.js' : `nuxt.config.${typescript ? 'ts' : 'js'}`
```

File: saofile.js [hunk 7]
```diff
@@ -9,7 +9,8 @@ module.exports = {
   prompts: require('./prompts'),
   templateData () {
     const typescript = this.answers.language.includes('ts')
-    const tsRuntime = this.answers.runtime && this.answers.runtime.includes('ts-runtime')
+    const tsRuntime =
+      this.answers.runtime && this.answers.runtime.includes('ts-runtime')
```
Comment (kevinmarrec) [inline saofile.js:13]: IMO Unnecessary change

File: saofile.js [hunk 8]
```diff
@@ -24,7 +25,18 @@ module.exports = {
     const { cliOptions = {} } = this.sao.opts
     const edge = cliOptions.edge ? '-edge' : ''
 
+    const ifTrue = (condition, content, elseContent) => condition ? content : elseContent || ''
+    const generateComponent = (imports, componentOptions) => {
+      return `
+<script ${ifTrue(typescript, 'lang="ts"')}>
+${ifTrue(typescript, 'import Vue from \'vue\'')}
+${ifTrue(!!imports, imports)}
```
Comment (kevinmarrec) [inline saofile.js:33]: I think that is there are imports, a carriage return should be ensured between imports and export.

**Comment by kevinmarrec** [comment]
@NickBolles Could you take into account my requested changes, then make your PR compatible with new monorepo (that's why you get lot of conflicts)

**Comment by kevinmarrec** [comment]
@NickBolles Can you fix conflicts & tests ? Thanks :)

**Comment by NickBolles** [comment]
@kevinmarrec updated. Could you take a look at this today? It's getting kinda difficult to keep this up to date with all of the other changes going on. :/

**Comment by kaangokdemir** [comment]
Greate feature and nice work @NickBolles

**Comment by Atinux** [comment]
@NickBolles fo you mind resolving the conflicts one last time?

**Review by lastthyme — APPROVED** [review]

**Comment by scscgit** [comment]
Is there any plan to continue working on this, or will it be changed after Nuxt 3 anyway?
